### PR TITLE
CLAP 1.2.1

### PR DESCRIPTION
**Pull request by abique**

### CODE REVIEW

**Comment by baconpaul** [comment]
The content is fine with me, but can we clean up the git transactions a bit? Like, rebase/merge all those undo changes into a single commit in the log?

**Comment by abique** [comment]
> The content is fine with me, but can we clean up the git transactions a bit? Like, rebase/merge all those undo changes into a single commit in the log?

We'd loose the history if we squash, also the next branch is on the repo and we can't force push, so I'm not sure if it'll help.

I value the history because we keep a record of the thinking process and the various steps the extension went through. It may be useful one day for someone trying to understand why we've made an extension in a certain way.

It is still possible to read a squashed change set by making a diff between tags.

File: include/clap/ext/draft/undo.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+///
+/// Calling host->undo() or host->redo() is equivalent to clicking undo/redo within the host's GUI.
+///
+/// If the plugin implements this interface then its undo and redo should be entirely delegated to
+/// the host; clicking in the plugin's UI undo or redo is equivalent to clicking undo or redo in the
+/// host's UI.
+///
+/// Some changes are long running changes, for example a mouse interaction will begin editing some
+/// complex data and it may take multiple events and a long duration to complete the change.
+/// In such case the plugin will call host->begin_change() to indicate the begining of a long
```
Comment (messmerd) [inline include/clap/ext/draft/undo.h:24]:
```suggestion
/// In such case the plugin will call host->begin_change() to indicate the beginning of a long
```
typo

File: include/clap/ext/draft/undo.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+};
+
+enum clap_undo_delta_properties_flags {
+   // If not set, then all clap_undo_delta_properties's attributes becomes irrelevant.
+   // If set, then the plugin will provide deltas in host->change_made().
+   CLAP_UNDO_DELTA_PROPERTIES_HAS_DELTA = 1 << 0,
+
+   // If set, then the delta will be re-usable in the future as long as the plugin is
+   // compatible with the given format_version.
+   CLAP_UNDO_DELTA_PROPERTIES_IS_PERSISTANT = 1 << 1,
```
Comment (messmerd) [inline include/clap/ext/draft/undo.h:56]:
```suggestion
   // If set, then the delta will be reusable in the future as long as the plugin is
   // compatible with the given format_version.
   CLAP_UNDO_DELTA_PROPERTIES_IS_PERSISTENT = 1 << 1,
```
typos

File: include/clap/ext/draft/undo.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+};
+
+typedef struct clap_undo_delta_properties {
+   // Bitmask of clap_undo_delta_properties_flags
+   uint64_t flags;
+
+   // This represent the delta format version that the plugin is using.
```
Comment (messmerd) [inline include/clap/ext/draft/undo.h:63]:
```suggestion
   // This represents the delta format version that the plugin is using.
```

File: include/clap/ext/draft/undo.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+   // cancel_change() must not be called without a preceding begin_change().
+   // [main-thread]
+   void(CLAP_ABI *cancel_change)(const clap_host_t *host);
+
+   // Completes an undoable change.
+   // At the moment of this function call, plugin_state->save() would include the current change.
+   //
+   // name: mandatory null terminated string describing the change, this is displayed to the user
+   //
+   // detlas: optional, they are binary blobs used to perform the undo and redo. When not available
```
Comment (messmerd) [inline include/clap/ext/draft/undo.h:115]:
```suggestion
   // deltas: optional, they are binary blobs used to perform the undo and redo. When not available
```
typo

**Comment by baconpaul** [comment]
We don’t have to squash. We can partially rebase next to squash some of the interim commits with rebase -I and the do pick and fixup

File: include/clap/ext/draft/undo.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+   // the host will save the plugin state and use state->load() to perform undo and redo.
+   //
+   // Note: the provided delta may be used for incremental state saving and crash recovery. The
+   // plugin can indicate a format version id and the validity lifetime for the binary blobs.
+   // The host can use to verify the compatiblilty before applying the delta.
+   // If the plugin is not able to use a delta, a notification should be produced to the user and
+   // the crash recovery will do a best effort job, at least restore the latest saved state.
```
Comment (messmerd) [inline include/clap/ext/draft/undo.h:122]:
```suggestion
   // The host can use these to verify the compatibility before applying the delta.
   // If the plugin is unable to use a delta, a notification should be provided to the user and
   // the crash recovery should perform a best effort job, at least restoring the latest saved state.
```
Typos, wording

**Review by messmerd — COMMENTED** [review]
Some suggestions - mostly typo fixes

**Comment by baconpaul** [comment]
Alex: I added info on rebasing on discord if that helps!

File: include/clap/ext/draft/undo.h [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+   // Subscribes or unsubscribe to undo context info.
+   // The plugin may only need the context info if it's GUI is shown and it wants to display
+   // undo/redo info.
+   //
+   // Initial state is unsubscribed.
+   //
+   // wants_info: set to true to receive context info
+   //
+   // [main-thread]
+   void(CLAP_ABI *set_wants_context_info)(const clap_host_t *host, bool wants_info);
```
[the remark this answers is not in the document]
Comment (abique) [reply include/clap/ext/draft/undo.h:155]: Let's go for `set_context_info_subscription`.
Thank you so much for all the changes! :+1:

**Review by Bremmers — APPROVED** [review]

**Review by defiantnerd — APPROVED** [review]
Looks fine to me. 2.0 ist much improved.

**Review by baconpaul — APPROVED** [review]
Yeah I'm happy with this!

**Pull request merged**